### PR TITLE
Cache file status in directory listing

### PR DESCRIPTION
**Pull request by qqibrow**
Introduce CacheDirectoryLister to cache all LocatedFileStatus using directory path as key

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        this.locatedFileStatusCache = CacheBuilder.newBuilder().maximumWeight(maxSize).weigher((Weigher<Path, List<LocatedFileStatus>>) (key, value) -> value.size()).expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS).recordStats().build();
+        this.tableNames = tables == null ? ImmutableSet.of() : tables.stream().map(CachedDirectoryLister::parseTableName).collect(Collectors.toSet());
+    }
+
+    private static SchemaTableName parseTableName(String tableName)
+    {
+        String[] parts = tableName.split("\\.");
+        if (parts.length != 2) {
+            throw new IllegalArgumentException("Invalid schemaTableName " + tableName);
+        }
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java]: How about using `checkArgument`?
```java
checkArgument(parts.length == 2, "Invalid schemaTableName: %s", tableName);
```
Ref: https://github.com/prestosql/presto/blob/57051c9a9616eee4f8713f58b84c07a8041767cb/presto-redis/src/main/java/io/prestosql/plugin/redis/RedisTableDescriptionSupplier.java#L119-L125

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java [hunk 2]
```diff
@@ -153,6 +155,10 @@
     private boolean isTemporaryStagingDirectoryEnabled = true;
     private String temporaryStagingDirectoryPath = "/tmp/presto-${USER}";
 
+    private Duration fileStatusCacheExpireAfterWrite = new Duration(1, TimeUnit.MINUTES);
+    private long fileStatusCacheMaxSize = 1000 * 1000;
+    private List<String> fileStatusCacheTables;
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java]: How about
```java
private List<String> fileStatusCacheTables = ImmutableList.of()
```

L966
```java
this.fileStatusCacheTables = fileStatusCacheTables == SPLITTER.splitToList(fileStatusCacheTables);
```

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    public CachedDirectoryLister(HiveClientConfig hiveClientConfig)
+    {
+        this(hiveClientConfig.getFileStatusCacheExpireAfterWrite(), hiveClientConfig.getFileStatusCacheMaxSize(), hiveClientConfig.getFileStatusCacheTables());
+    }
+
+    private CachedDirectoryLister(Duration expireAfterWrite, long maxSize, List<String> tables)
+    {
+        this.locatedFileStatusCache = CacheBuilder.newBuilder().maximumWeight(maxSize).weigher((Weigher<Path, List<LocatedFileStatus>>) (key, value) -> value.size()).expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS).recordStats().build();
+        this.tableNames = tables == null ? ImmutableSet.of() : tables.stream().map(CachedDirectoryLister::parseTableName).collect(Collectors.toSet());
+    }
```
Comment (ebyhr) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java]: I think these methods should have new lines for the readbility.
```java
        this.locatedFileStatusCache = CacheBuilder.newBuilder()
                                                  .maximumWeight(maxSize)
                                                  .weigher((Weigher<Path, List<LocatedFileStatus>>) (key, value) -> value.size())
                                                  .expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS)
                                                  .recordStats()
                                                  .build();
        this.tableNames = tables.stream()
                                .map(CachedDirectoryLister::parseTableName)
                                .collect(Collectors.toSet());
```

**Comment by ebyhr** [comment]
Let's capitalize the commit message. Also, I feel`cache located file status` is not clear (`hive` or `hdfs` should be included?) because this is connector specific change.
https://chris.beams.io/posts/git-commit/#capitalize

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        }
+        return new SchemaTableName(parts[0], parts[1]);
+    }
+
+    @Override
+    public RemoteIterator<LocatedFileStatus> list(FileSystem fs, Table table, Path path)
+            throws IOException
+    {
+        List<LocatedFileStatus> files = locatedFileStatusCache.getIfPresent(path);
+        if (files != null) {
```
Comment (luohao) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java]: When will the cached entry be evicted/invalidated?

If there's a change to the path (e.g., a new file added to a partition) and that path always hit cache in `list()` operation, it seems to me that Presto will never be able to see the newly added files.

**Comment by electrum** [comment]
Add a `@Managed` method `flushCache()` to allow clearing the cache if needed via JMX.

**Comment by electrum** [comment]
@qqibrow Thanks for contributing this. Please ping me after you have addressed the comments so that we can get it merged.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java (hunk 1 above)
Comment (qqibrow) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java]: sounds good.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java (hunk 4 above)
Comment (qqibrow) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/CachedDirectoryLister.java]: This cache works best when your ingestion pattern is immutable, e,g, insert new partition to a table. If, insert overwrite a table, the cache could lead to query failure if timeout is longer than ingestion frequency. That's why I added a list of table to cache in config.

I was thinking reader error could feedback cache here to evict stale entries. However, I didn't think of good solution.

**Comment by qqibrow** [comment]
> Add a `@Managed` method `flushCache()` to allow clearing the cache if needed via JMX.

Nice suggestion! I added:
```
    @Managed
    public void flushCache()
    {
        cache.invalidateAll();
    }
```
Is it possible to trigger than through` /v1/jmx/...` endpoint as well?

**Comment by electrum** [comment]
The HTTP JMX endpoint currently only supports reads. It's possible to extend it to support writes, but that would have security implications. For now, you need to use a JMX client.

Another option is to add a callable procedure that flushes the cache, though that won't work when we support multiple coordinators.

**Comment by electrum** [comment]
Merged, thanks!

**Pull request merged**